octodns/provider/selectel.py

```diff
@@ -0,0 +1,278 @@
+        values = []
+        for record in records:
+            values.append({
+                'priority': record['priority'],
+                'weight': record['weight'],
+                'port': record['port'],
+                'target': '{}.'.format(record['target']),
+            })
+
+        return {
+            'type': _type,
+            'ttl': records[0]['ttl'],
+            'values': values,
+        }
+
+    def populate(self, zone, target=False, lenient=False):
+        self.log.debug('populate: name=%s, target=%s, lenient=%s',
+                       zone.name, target, lenient)
+        before = len(zone.records)
+        records = self.zone_records(zone)
+        if records:
+            values = defaultdict(lambda: defaultdict(list))
+            for record in records:
+                name = zone.hostname_from_fqdn(record['name'])
+                _type = record['type']
+                if _type in self.SUPPORTS:
+                    values[name][record['type']].append(record)
+            for name, types in values.items():
+                for _type, records in types.items():
+                    data_for = getattr(self, '_data_for_{}'.format(_type))
+                    data = data_for(_type, records)
+                    record = Record.new(zone, name, data, source=self,
+                                        lenient=lenient)
+                    zone.add_record(record)
+        self.log.info('populate:   found %s records',
+                      len(zone.records) - before)
+
+    def domain_list(self):
+        path = '/'
+
+        resp = self._request('GET', path)
+        domain_dict = {}
+
+        for domain in resp:
+            domain_dict[domain['name']] = domain
+
+        return domain_dict
+
+    def zone_records(self, zone):
+        path = '/{}/records/'.format(zone.name[:-1])
+        resp = self._request('GET', path)
+
+        self._zone_records[zone.name] = resp
+
+        return self._zone_records[zone.name]
+
+    def create_domain(self, name, zone=""):
+        path = '/'
+
+        data = {
+            'name': name,
+            'bind_zone': zone,
+        }
+
+        resp = self._request('POST', path, data=data)
+        self._domain_list[name] = resp
+        return resp
+
+    def create_record(self, zone_name, data):
+        self.log.debug('Create record. Zone: %s, data %s', zone_name, data)
+        if zone_name in self._domain_list.keys():
+            domain_id = self._domain_list[zone_name]['id']
+        else:
+            domain_id = self.create_domain(zone_name)['id']
+
+        path = '/{}/records/'.format(domain_id)
+        return self._request('POST', path, data=data)
+
+    def delete_record(self, domain, _type, zone):
+        self.log.debug('Delete record. Domain: %s, Type: %s', domain, _type)
+
+        domain_id = self._domain_list[domain]['id']
+        path = '/{}/records/'.format(domain_id)
+
+        records = self._request('GET', path)
```

Comment:
This would be O(n^2) for deleting records which would likely run into trouble when deleting many records from a large zone. It may make sense to look at the caching mechanism that some of the other providers incorporate to avoid this problem.

Reply:
Something like this bit here: https://github.com/github/octodns/blob/master/octodns/provider/cloudflare.py#L71-L87

`self.zones.get(domain, False)` is called instead of doing an API call directly.